src/software/simulation/physics_simulator_robot.h

```diff
@@ -0,0 +1,311 @@
+
+   private:
+    /**
+     * A function that is called during every physics step for as long as the ball
+     * is touching this robot's dribbler
+     *
+     * @param physics_robot The robot involved in the contact
+     * @param physics_ball The ball invovled in the contact
+     */
+    void onDribblerBallContact(PhysicsRobot* physics_robot, PhysicsBall* physics_ball);
+
+    /**
+     * A function that is called during once when the ball starts touching
+     * this robot's dribbler.
+     *
+     * @param physics_robot The robot involved in the contact
+     * @param physics_ball The ball invovled in the contact
+     */
+    void onDribblerBallStartContact(PhysicsRobot* physics_robot,
+                                    PhysicsBall* physics_ball);
+
+    /**
+     * A function that is called during once when the ball stops touching
+     * this robot's dribbler.
+     *
+     * @param physics_robot The robot involved in the contact
+     * @param physics_ball The ball invovled in the contact
+     */
+    void onDribblerBallEndContact(PhysicsRobot* physics_robot, PhysicsBall* physics_ball);
+
+    /**
+     * Helper functions that check if the current pointer to the physics_robot is valid
+     * before calling the given function. If the physics_robot is invalid, a warning is
+     * logged and a default value is returned.
+     *
+     * @param func The function to perform on the physics robot
+     */
+    void checkValidAndExecuteVoid(
+        std::function<void(std::shared_ptr<PhysicsRobot>)> func);
+    float checkValidAndReturnFloat(
+        std::function<float(std::shared_ptr<PhysicsRobot>)> func);
+    unsigned int checkValidAndReturnUint(
+        std::function<unsigned int(std::shared_ptr<PhysicsRobot>)> func);
+
+    /**
+     * Applies force to the physics ball to simulate it being dribbled by the
+     * physics robot.
+     *
+     * @param physics_robot The robot that should dribble the ball
+     * @param physics_ball The ball to be dribbled
+     */
+    void applyDribblerForce(PhysicsRobot* physics_robot, PhysicsBall* physics_ball);
+
+    std::weak_ptr<PhysicsRobot> physics_robot;
+    std::optional<float> autokick_speed_m_per_s;
+    std::optional<float> autochip_distance_m;
+    uint32_t dribbler_rpm;
+
+    typedef struct DribblerBall_t
+    {
+        PhysicsBall* ball;
+        // We keep track of whether or not the ball can be controlled
+        // in any way by the robot. This includes kicking, chipping,
+        // and dribbling. This extra information is used to
+        // prevent edge cases like the ball getting kicked/chipped
+        // multiple times, and to prevent the dribbler from affecting
+        // kicking
+        bool can_be_controlled;
+    } DribblerBall;
+
+    std::optional<DribblerBall> ball_in_dribbler_area;
+
+    std::unique_ptr<PrimitiveManager, FirmwarePrimitiveManagerDeleter> primitive_manager;
```

Comment:
would this be worth putting in the base class?

Reply:
Don't think it makes much of a difference, would just have to make it protected or public. Let me know if you prefer it in the base class

Reply:
Moved to base class